src/nsrt_learning/side_predicate_learning.py

```diff
@@ -456,18 +495,68 @@ def _sideline(self) -> List[PartialNSRTAndDatastore]:
+        # make a list of all final PNADs. Note
+        # that these final PNADs only come from the
+        # param_opt_to_nec_pnad dict, since we can be assured
+        # that our backchaining process ensured that the
+        # PNADs in this dict cover all of the data!
+        all_pnads = []
+        for pnad_list in param_opt_to_nec_pnad.values():
+            for pnad in pnad_list:
+                all_pnads.append(pnad)
+
         # At this point, all PNADs have correct parameters, preconditions,
         # and add effects. We now finalize the delete effects and side
         # predicates. Note that we have to do delete effects first, and
         # then side predicates, because the latter rely on the former.
-        for pnad in param_opt_to_pnad.values():
+        print()
+        for pnad in all_pnads:
             self._finalize_pnad_delete_effects(pnad)
             self._finalize_pnad_side_predicates(pnad)
+            print(pnad)
 
         # Before returning, sanity check that harmlessness holds.
-        final_pnads = list(param_opt_to_pnad.values())
-        assert self._check_harmlessness(final_pnads)
-        return final_pnads
+        assert self._check_harmlessness(all_pnads)
+        return all_pnads
+
+    @staticmethod
+    def _find_unification(
+        necessary_add_effects: Set[GroundAtom],
+        pnad: PartialNSRTAndDatastore,
+        segment: Segment,
+        find_partial_grounding: bool = False
+    ) -> Optional[_GroundSTRIPSOperator]:
+        # Find a mapping from the variables in the PNAD add effects
+        # and option to the objects in necessary_add_effects and the
+        # segment's option. If one exists, we don't need to modify this
+        # PNAD. Otherwise, we must make its add effects more specific.
+        # Note that we are assuming all variables in the parameters
+        # of the PNAD will appear in either the option arguments or
+        # the add effects. This is in contrast to strips_learning.py,
+        # where delete effect variables also contribute to parameters.
+        objects = list(segment.states[0])
+        option_objs = segment.get_option().objects
+        isub = dict(zip(pnad.option_spec[1], option_objs))
+        # Loop over all groundings.
+        for ground_op in utils.all_ground_operators_given_partial(
+                pnad.op, objects, isub):
+            if not ground_op.preconditions.issubset(segment.init_atoms):
+                continue
+            # If find_partial_grounding is True, we want to find a grounding
+            # that achieves some subset of the necessary add effects. Else,
+            # we want to find a grounding that is some superset of the
+            # necessary_add_effects
+            if find_partial_grounding:
+                if not ground_op.add_effects.issubset(necessary_add_effects):
+                    continue
+            else:
+                if not ground_op.add_effects.issubset(segment.final_atoms):
```

Comment:
shouldn't this `if` clause be outside the `else:` right above? you suggested that to me haha i had it this way before

Reply:
no - i think this is right; we only want to run this check if we're not trying to find a partial grounding

Reply:
gotcha -- note that this is different from before. it's okay though